### ISSUE TITLE
Clicking on Qtox Icon on desktop tries to open separate instance

### ISSUE THREAD

**Issue by DeerSpotter** [qTox/qTox#2917]
Clicking on the icon on desktop shortcut tries to open up a separate instance and it doesn't do it. loading the same profile fails. Which is good. But we would like it to just quickly open a opened program instead. 

Any way of actually opening the program just as if you clicked it in the notification task bar?

**Comment by zetok** [comment]
Your OS, etc?
https://github.com/tux3/qTox/blob/master/CONTRIBUTING.md#must-read

**Comment by TheNain38** [comment]
@zetok He want a single instance mode in qTox, because for now, if you reopen qTox, it will open another instance. You would need to add another setting...

**Comment by ProMcTagonist** [comment]
~~iirc long term plan is to remove multiple instances and handle multiple profiles online at once inside qTox, but don't quote me until i can cite an issue number and comment~~

**Comment by zetok** [comment]
@ProMcTagonist 

> iirc long term plan is to remove multiple instances and handle multiple profiles online at once inside qTox, but don't quote me until i can cite an issue number and comment

Err, not really, it was decided that doing multiple IDs would be too complex & complicated, and thus making multiple instances supported.

**Comment by ProMcTagonist** [comment]
yup, not sure where I remembered that from, I've found tux3 detailing why it's impossible hell in #1018 so nevermind

**Comment by ghost** [comment]
> desktop shortcut
> 
> > Your OS, etc?

Sounds like Windows. I can reproduce in Win XP x32 and Win 7 x32 boxes.

---

Happens with both portable and installed qTox. The problem relies in the exe, not the shortcut.

**Comment by DeerSpotter** [comment]
here is a C# equivalent of opening a instance that is open already, and if not open a new instance.


```
private const  int SW_SHOWNORMAL = 1;

[DllImport("user32.dll", SetLastError = true, CharSet = CharSet.Auto)]
private static extern bool ShowWindow(IntPtr hwnd, int nCmdShow);


[DllImport("user32.dll", SetLastError = true)]
private static extern bool SetForegroundWindow(IntPtr hwnd);

public void SetForeground()
{
    Process[] processes = Process.GetProcessesByName("process name");

    foreach (Process p in processes) {
        ShowWindow(p.MainWindowHandle, SW_SHOWNORMAL);
        SetForegroundWindow(p.MainWindowHandle);
    }
}
```

**Comment by DeerSpotter** [comment]
found out that creating a shortcut like so: "C:\Program Files\qTox\bin\qtox.exe" -p "tox_save"
auto opens the profile (in my case called: "tox save")

**Comment by sudden6** [comment]
This is weird, on linux it opens the active instance.
Please provide the console output of both instances.

**Comment by DeerSpotter** [comment]
I wont be able to give you console output as my "lead" told me to remove qtox from all the test machines. Now we are going the office 365 path :/

**Comment by aphirst** [comment]
I'm having this issue too in Windows 7 with the (currently) most recent version of qTox on my parents' machine, and combined with the various close/minimise-to-tray options (either enabled or disabled) the resulting confusion is too much for them to bear.

(I want their logged-in session to persist them minimising/closing the Window, but also for them to be able to "re-open" that instance by clicking a Pinned qTox icon from the taskbar, since they have serious issues comprehending the System Tray.)

**Comment by anthonybilinski** [comment]
This is still reproducible on v1.17.2.

Looking into the logs, we are attaching to global IPC correctly:
```
} else if (globalMemory.attach()) {
    qDebug() << "Attaching to the global shared memory";
```
```
ipc.cpp:73 : Debug: Attaching to the global shared memory
```
and we must be parsing arguments correctly, otherwise we would exit:
```
if (parser.positionalArguments().empty()) {
    eventType = "activate";
} else {
    firstParam = parser.positionalArguments()[0];
    // Tox URIs. If there's already another qTox instance running, we ask it to handle the URI
    // and we exit
    // Otherwise we start a new qTox instance and process it ourselves
    if (firstParam.startsWith("tox:")) {
        eventType = "uri";
    } else if (firstParam.endsWith(".tox")) {
        eventType = "save";
    } else {
        qCritical() << "Invalid argument";
        return EXIT_FAILURE;
    }
}
```
we means we should be sending out the IPC call. But we don't do that if we're the IPC owner:
```
if (doIpc && !ipc.isCurrentOwner()) {
```
which is defined by whether the IPC's owner ID matches our `globalId`. Well, it looks like on Windows, every instance of qTox has exactly the same globalId of 3169903406987816468:
```
ipc.cpp:61 : Debug: Our global IPC ID is  3169903406987816468
```
This can even be seen in logs from a few different qTox tickets from a search for that number. On Linux, that ID is different on each run.

We're generating the ID with:
```
std::default_random_engine randEngine((std::random_device())());
std::uniform_int_distribution<uint64_t> distribution;
globalId = distribution(randEngine);
```
so that's what needs to be fixed to fix this issue on Windows.

**Comment by anthonybilinski** [comment]
https://en.cppreference.com/w/cpp/numeric/random/random_device
> A notable implementation where std::random_device is deterministic is old versions of MinGW (bug 338, fixed since GCC 9.2)

Changing this to just use time as the seed rather than a true random device, although it should be way less random, should be good enough for our IPC ID, and continue to work with mingw, without having to add platform specific code.
